File: app/src/test/java/com/mss/imagesearcher/view/imagelist/MainActivityTest.java
package com.mss.imagesearcher.view.imagelist;

import android.content.Intent;
import android.view.KeyEvent;
import android.view.View;
import android.widget.ProgressBar;

import androidx.recyclerview.widget.RecyclerView;
import androidx.swiperefreshlayout.widget.SwipeRefreshLayout;

import com.google.android.material.textfield.TextInputEditText;
import com.mss.imagesearcher.R;
import com.mss.imagesearcher.di.RobolectricApp;
import com.mss.imagesearcher.di.RobolectricComponent;
import com.mss.imagesearcher.domain.models.ImageInfo;
import com.mss.imagesearcher.presenter.imagelist.ImageListPresenter;
import com.mss.imagesearcher.robolectric.ShadowSnackbar;
import com.mss.imagesearcher.view.detail.DetailActivity;
import com.mss.imagesearcher.view.helpers.ImageLoader;
import com.mss.imagesearcher.view.main.MainActivity;

import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.robolectric.Robolectric;
import org.robolectric.RobolectricTestRunner;
import org.robolectric.Shadows;
import org.robolectric.annotation.Config;
import org.robolectric.shadows.ShadowApplication;

import java.util.ArrayList;
import java.util.List;

import javax.inject.Inject;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertFalse;
import static org.junit.Assert.assertTrue;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.Mockito.doAnswer;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

@RunWith(RobolectricTestRunner.class)
@Config(sdk = 28, application = RobolectricApp.class, shadows = {ShadowSnackbar.class})
public class MainActivityTest {

    private MainActivity mainActivity;
    private List<ImageInfo> imageInfoList;

    @Inject
    ImageListPresenter presenter;

    @Inject
    ImageLoader imageLoader;

    @Before
    public void setUp() {
        mainActivity = Robolectric.buildActivity(MainActivity.class)
                .create()
                .start()
                .resume()
                .visible()
                .get();
        ((RobolectricComponent) RobolectricApp.getAppComponent()).inject(this);
    }

    @Test
    public void showProgress_Hide_pbDetailHide() {
        ProgressBar progressBar = mainActivity.findViewById(R.id.pbList);

        mainActivity.showProgress(false);

        assertEquals(View.INVISIBLE, progressBar.getVisibility());
    }

    @Test
    public void showProgress_Show_pbDetailShow() {
        ProgressBar progressBar = mainActivity.findViewById(R.id.pbList);

        mainActivity.showProgress(true);

        assertEquals(View.VISIBLE, progressBar.getVisibility());
    }

    @Test
    public void stopRefreshing_pbDetailShow() {
        SwipeRefreshLayout swipeRefreshLayout = mainActivity.findViewById(R.id.srlImages);
        swipeRefreshLayout.setRefreshing(true);

        mainActivity.stopRefreshing();

        assertFalse(swipeRefreshLayout.isRefreshing());
    }

    @Test
    public void refreshImageList_callPresenterGetRvPresenter() {
        initImageInfoList();
        ImageListPresenter.RvPresenter rvPresenter = mock(ImageListPresenter.RvPresenter.class);
        when(presenter.getRvPresenter()).thenReturn(rvPresenter);
        when(rvPresenter.getItemCount()).thenReturn(imageInfoList.size());

        mainActivity.refreshImageList();
        mainActivity.refreshImageList();

        verify(presenter).getRvPresenter();
    }

    @Test
    public void showNotFoundMessage_showSnackbar() {
        ShadowSnackbar.reset();

        mainActivity.showNotFoundMessage();

        assertEquals(ShadowSnackbar.getTextOfLatestSnackbar(), mainActivity.getString(R.string.not_found_message));
    }

    @Test
    public void openDetailActivity_startDetailActivityWithParam() {
        mainActivity.openDetailActivity(1);

        Intent intent = Shadows.shadowOf(mainActivity).getNextStartedActivity();
        assertEquals("com.mss.imagesearcher.view.detail.DetailActivity", intent.getComponent().getClassName());
        assertEquals(1, intent.getIntExtra(DetailActivity.PARAMETER_POSITION_TAG, 0));
    }

    @Test
    public void onRequestPermissionsResult_NoPermission_FinishActivity() {
        mainActivity.onRequestPermissionsResult(MainActivity.PERMISSION_REQUEST_CODE, new String[2], new int[2]);
        assertTrue(Shadows.shadowOf(mainActivity).isFinishing());
    }

    @Test
    public void onRequestPermissionsResult_HasPermission_NoFinishActivity() {
        ShadowApplication application = new ShadowApplication();
        for (String p : MainActivity.Companion.getNETWORK_PERMISSIONS()) {
            application.grantPermissions(p);
        }

        mainActivity.onRequestPermissionsResult(MainActivity.PERMISSION_REQUEST_CODE, new String[2], new int[2]);

        assertTrue(!Shadows.shadowOf(mainActivity).isFinishing());
    }

    @Test
    public void onItemClick_callPresenter() {
        initImageInfoList();
        ImageListPresenter.RvPresenter rvPresenter = mock(ImageListPresenter.RvPresenter.class);
        when(presenter.getRvPresenter()).thenReturn(rvPresenter);
        when(rvPresenter.getItemCount()).thenReturn(imageInfoList.size());
        RecyclerView recyclerView = mainActivity.findViewById(R.id.rvImages);
        recyclerView.measure(0, 0);
        recyclerView.layout(0, 0, 100, 800);
        mainActivity.refreshImageList();
        ImageListAdapter.ViewHolder viewHolder = (ImageListAdapter.ViewHolder) recyclerView.findViewHolderForAdapterPosition(1);

        viewHolder.getIvItem().performClick();

        verify(presenter).onItemClick(1);
    }

    @Test
    public void holderSetImage_SuccessLoad_callPresenterOnImageLoaded() {
        initImageInfoList();
        ImageListPresenter.RvPresenter rvPresenter = mock(ImageListPresenter.RvPresenter.class);
        when(presenter.getRvPresenter()).thenReturn(rvPresenter);
        when(rvPresenter.getItemCount()).thenReturn(imageInfoList.size());
        RecyclerView recyclerView = mainActivity.findViewById(R.id.rvImages);
        recyclerView.measure(0, 0);
        recyclerView.layout(0, 0, 100, 200);
        mainActivity.refreshImageList();
        ImageListAdapter.ViewHolder viewHolder = (ImageListAdapter.ViewHolder) recyclerView.findViewHolderForAdapterPosition(1);
        doAnswer(answer -> {
            ((Runnable) answer.getArgument(3)).run();
            return null;
        }).when(imageLoader).load(any(), any(), any(), any(), any());

        viewHolder.setImage("path");

        assertEquals(1, viewHolder.getPos());
        verify(rvPresenter).onImageLoaded(any());
    }

    @Test
    public void holderSetImage_FailLoad_callPresenterOnImageLoaded() {
        initImageInfoList();
        ImageListPresenter.RvPresenter rvPresenter = mock(ImageListPresenter.RvPresenter.class);
        when(presenter.getRvPresenter()).thenReturn(rvPresenter);
        when(rvPresenter.getItemCount()).thenReturn(imageInfoList.size());
        RecyclerView recyclerView = mainActivity.findViewById(R.id.rvImages);
        recyclerView.measure(0, 0);
        recyclerView.layout(0, 0, 100, 200);
        mainActivity.refreshImageList();
        ImageListAdapter.ViewHolder viewHolder = (ImageListAdapter.ViewHolder) recyclerView.findViewHolderForAdapterPosition(1);
        doAnswer(answer -> {
            ((Runnable) answer.getArgument(4)).run();
            return null;
        }).when(imageLoader).load(any(), any(), any(), any(), any());

        viewHolder.setImage("path");

        verify(rvPresenter).onImageLoadFailed(any());
    }

    @Test
    public void holderShowProgress_True_ProgressBarVisible() {
        initImageInfoList();
        ImageListPresenter.RvPresenter rvPresenter = mock(ImageListPresenter.RvPresenter.class);
        when(presenter.getRvPresenter()).thenReturn(rvPresenter);
        when(rvPresenter.getItemCount()).thenReturn(imageInfoList.size());
        RecyclerView recyclerView = mainActivity.findViewById(R.id.rvImages);
        recyclerView.measure(0, 0);
        recyclerView.layout(0, 0, 100, 200);
        mainActivity.refreshImageList();
        ImageListAdapter.ViewHolder viewHolder = (ImageListAdapter.ViewHolder) recyclerView.findViewHolderForAdapterPosition(1);

        viewHolder.showProgress(true);

        assertEquals(View.VISIBLE, viewHolder.getPbItem().getVisibility());
    }

    @Test
    public void holderShowProgress_False_ProgressBarVisible() {
        initImageInfoList();
        ImageListPresenter.RvPresenter rvPresenter = mock(ImageListPresenter.RvPresenter.class);
        when(presenter.getRvPresenter()).thenReturn(rvPresenter);
        when(rvPresenter.getItemCount()).thenReturn(imageInfoList.size());
        RecyclerView recyclerView = mainActivity.findViewById(R.id.rvImages);
        recyclerView.measure(0, 0);
        recyclerView.layout(0, 0, 100, 200);
        mainActivity.refreshImageList();
        ImageListAdapter.ViewHolder viewHolder = (ImageListAdapter.ViewHolder) recyclerView.findViewHolderForAdapterPosition(1);

        viewHolder.showProgress(false);

        assertEquals(View.GONE, viewHolder.getPbItem().getVisibility());
    }

    @Test
    public void onAction_callPresenterOnSearchClick() {
        TextInputEditText etSearch = mainActivity.findViewById(R.id.etSearch);
        etSearch.setText("one");
        etSearch.onKeyDown(KeyEvent.KEYCODE_ENTER, new KeyEvent(KeyEvent.ACTION_DOWN, KeyEvent.KEYCODE_ENTER));

        verify(presenter).onSearchClick("one");
    }

    private void initImageInfoList() {
        imageInfoList = new ArrayList();
        for (int i = 0; i < 50; i++) {
            ImageInfo imageInfo = new ImageInfo();
            imageInfo.setId(i);
            imageInfo.setPreviewURL("https:\\\\previewurl" + i);
            imageInfo.setLargeImageURL("https:\\\\largeimagewurl" + i);
            imageInfoList.add(imageInfo);
        }
    }
}

File: app/src/test/java/com/mss/imagesearcher/view/imagelist/MainActivityTest.kt
package com.mss.imagesearcher.view.imagelist

import android.view.KeyEvent
import android.view.View
import android.widget.ProgressBar
import androidx.recyclerview.widget.RecyclerView
import androidx.swiperefreshlayout.widget.SwipeRefreshLayout
import com.google.android.material.textfield.TextInputEditText
import com.mss.imagesearcher.R
import com.mss.imagesearcher.di.RobolectricApp
import com.mss.imagesearcher.di.RobolectricComponent
import com.mss.imagesearcher.domain.models.ImageInfo
import com.mss.imagesearcher.presenter.imagelist.ImageListPresenter
import com.mss.imagesearcher.robolectric.ShadowSnackbar
import com.mss.imagesearcher.view.detail.DetailActivity
import com.mss.imagesearcher.view.helpers.ImageLoader
import com.mss.imagesearcher.view.main.MainActivity
import org.junit.Assert.*
import org.junit.Before
import org.junit.Test
import org.junit.runner.RunWith
import org.mockito.ArgumentMatchers.any
import org.mockito.Mockito.*
import org.robolectric.Robolectric
import org.robolectric.RobolectricTestRunner
import org.robolectric.Shadows
import org.robolectric.annotation.Config
import org.robolectric.shadows.ShadowApplication
import java.util.*
import javax.inject.Inject

@RunWith(RobolectricTestRunner::class)
@Config(sdk = [28], application = RobolectricApp::class, shadows = [ShadowSnackbar::class])
class MainActivityTest {

    private var mainActivity: MainActivity? = null
    private var imageInfoList: MutableList<ImageInfo>? = null

    @Inject
    var presenter: ImageListPresenter? = null

    @Inject
    var imageLoader: ImageLoader? = null

    @Before
    fun setUp() {
        mainActivity = Robolectric.buildActivity(MainActivity::class.java)
                .create()
                .start()
                .resume()
                .visible()
                .get()
        (RobolectricApp.appComponent as RobolectricComponent).inject(this)
    }

    @Test
    fun showProgress_Hide_pbDetailHide() {
        val progressBar = mainActivity!!.findViewById<ProgressBar>(R.id.pbList)

        mainActivity!!.showProgress(false)

        assertEquals(View.INVISIBLE.toLong(), progressBar.visibility.toLong())
    }

    @Test
    fun showProgress_Show_pbDetailShow() {
        val progressBar = mainActivity!!.findViewById<ProgressBar>(R.id.pbList)

        mainActivity!!.showProgress(true)

        assertEquals(View.VISIBLE.toLong(), progressBar.visibility.toLong())
    }

    @Test
    fun stopRefreshing_pbDetailShow() {
        val swipeRefreshLayout = mainActivity!!.findViewById<SwipeRefreshLayout>(R.id.srlImages)
        swipeRefreshLayout.isRefreshing = true

        mainActivity!!.stopRefreshing()

        assertFalse(swipeRefreshLayout.isRefreshing)
    }

    @Test
    fun refreshImageList_callPresenterGetRvPresenter() {
        initImageInfoList()
        val rvPresenter = mock(ImageListPresenter.RvPresenter::class.java)
        `when`<ImageListPresenter.RvPresenter>(presenter!!.rvPresenter).thenReturn(rvPresenter)
        `when`(rvPresenter.itemCount).thenReturn(imageInfoList!!.size)

        mainActivity!!.refreshImageList()
        mainActivity!!.refreshImageList()

        verify<ImageListPresenter>(presenter).rvPresenter
    }

    @Test
    fun showNotFoundMessage_showSnackbar() {
        ShadowSnackbar.reset()

        mainActivity!!.showNotFoundMessage()

        assertEquals(ShadowSnackbar.textOfLatestSnackbar, mainActivity!!.getString(R.string.not_found_message))
    }

    @Test
    fun openDetailActivity_startDetailActivityWithParam() {
        mainActivity!!.openDetailActivity(1)

        val intent = Shadows.shadowOf(mainActivity).nextStartedActivity
        assertEquals("com.mss.imagesearcher.view.detail.DetailActivity", intent.component!!.className)
        assertEquals(1, intent.getIntExtra(DetailActivity.PARAMETER_POSITION_TAG, 0).toLong())
    }

    @Test
    fun onRequestPermissionsResult_NoPermission_FinishActivity() {
        mainActivity!!.onRequestPermissionsResult(MainActivity.PERMISSION_REQUEST_CODE, arrayOfNulls(2), IntArray(2))
        assertTrue(Shadows.shadowOf(mainActivity).isFinishing)
    }

    @Test
    fun onRequestPermissionsResult_HasPermission_NoFinishActivity() {
        val application = ShadowApplication()
        for (p in MainActivity.NETWORK_PERMISSIONS) {
            application.grantPermissions(p)
        }

        mainActivity!!.onRequestPermissionsResult(MainActivity.PERMISSION_REQUEST_CODE, arrayOfNulls(2), IntArray(2))

        assertTrue(!Shadows.shadowOf(mainActivity).isFinishing)
    }

    @Test
    fun onItemClick_callPresenter() {
        initImageInfoList()
        val rvPresenter = mock(ImageListPresenter.RvPresenter::class.java)
        `when`<RvPresenter>(presenter!!.rvPresenter).thenReturn(rvPresenter)
        `when`(rvPresenter.itemCount).thenReturn(imageInfoList!!.size)
        val recyclerView = mainActivity!!.findViewById<RecyclerView>(R.id.rvImages)
        recyclerView.measure(0, 0)
        recyclerView.layout(0, 0, 100, 800)
        mainActivity!!.refreshImageList()
        val viewHolder = recyclerView.findViewHolderForAdapterPosition(1) as ImageListAdapter.ViewHolder?

        viewHolder!!.ivItem!!.performClick()

        verify<ImageListPresenter>(presenter).onItemClick(1)
    }

    @Test
    fun holderSetImage_SuccessLoad_callPresenterOnImageLoaded() {
        initImageInfoList()
        val rvPresenter = mock(ImageListPresenter.RvPresenter::class.java)
        `when`<RvPresenter>(presenter!!.rvPresenter).thenReturn(rvPresenter)
        `when`(rvPresenter.itemCount).thenReturn(imageInfoList!!.size)
        val recyclerView = mainActivity!!.findViewById<RecyclerView>(R.id.rvImages)
        recyclerView.measure(0, 0)
        recyclerView.layout(0, 0, 100, 200)
        mainActivity!!.refreshImageList()
        val viewHolder = recyclerView.findViewHolderForAdapterPosition(1) as ImageListAdapter.ViewHolder?
        doAnswer { answer ->
            (answer.getArgument<Any>(3) as Runnable).run()
            null
        }.`when`<ImageLoader>(imageLoader).load(any<Context>(), any(), any<ImageView>(), any(), any())

        viewHolder!!.setImage("path")

        assertEquals(1, viewHolder.pos.toLong())
        verify(rvPresenter).onImageLoaded(any())
    }

    @Test
    fun holderSetImage_FailLoad_callPresenterOnImageLoaded() {
        initImageInfoList()
        val rvPresenter = mock(ImageListPresenter.RvPresenter::class.java)
        `when`<RvPresenter>(presenter!!.rvPresenter).thenReturn(rvPresenter)
        `when`(rvPresenter.itemCount).thenReturn(imageInfoList!!.size)
        val recyclerView = mainActivity!!.findViewById<RecyclerView>(R.id.rvImages)
        recyclerView.measure(0, 0)
        recyclerView.layout(0, 0, 100, 200)
        mainActivity!!.refreshImageList()
        val viewHolder = recyclerView.findViewHolderForAdapterPosition(1) as ImageListAdapter.ViewHolder?
        doAnswer { answer ->
            (answer.getArgument<Any>(4) as Runnable).run()
            null
        }.`when`<ImageLoader>(imageLoader).load(any<Context>(), any(), any<ImageView>(), any(), any())

        viewHolder!!.setImage("path")

        verify<RvPresenter>(rvPresenter).onImageLoadFailed(any<IImageListViewHolder>())
    }

    @Test
    fun holderShowProgress_True_ProgressBarVisible() {
        initImageInfoList()
        val rvPresenter = mock(ImageListPresenter.RvPresenter::class.java)
        `when`<RvPresenter>(presenter!!.rvPresenter).thenReturn(rvPresenter)
        `when`(rvPresenter.itemCount).thenReturn(imageInfoList!!.size)
        val recyclerView = mainActivity!!.findViewById<RecyclerView>(R.id.rvImages)
        recyclerView.measure(0, 0)
        recyclerView.layout(0, 0, 100, 200)
        mainActivity!!.refreshImageList()
        val viewHolder = recyclerView.findViewHolderForAdapterPosition(1) as ImageListAdapter.ViewHolder?

        viewHolder!!.showProgress(true)

        assertEquals(View.VISIBLE.toLong(), viewHolder.pbItem!!.visibility.toLong())
    }

    @Test
    fun holderShowProgress_False_ProgressBarVisible() {
        initImageInfoList()
        val rvPresenter = mock(ImageListPresenter.RvPresenter::class.java)
        `when`<RvPresenter>(presenter!!.rvPresenter).thenReturn(rvPresenter)
        `when`(rvPresenter.itemCount).thenReturn(imageInfoList!!.size)
        val recyclerView = mainActivity!!.findViewById<RecyclerView>(R.id.rvImages)
        recyclerView.measure(0, 0)
        recyclerView.layout(0, 0, 100, 200)
        mainActivity!!.refreshImageList()
        val viewHolder = recyclerView.findViewHolderForAdapterPosition(1) as ImageListAdapter.ViewHolder?

        viewHolder!!.showProgress(false)

        assertEquals(View.GONE.toLong(), viewHolder.pbItem!!.visibility.toLong())
    }

    @Test
    fun onAction_callPresenterOnSearchClick() {
        val etSearch = mainActivity!!.findViewById<TextInputEditText>(R.id.etSearch)
        etSearch.setText("one")
        etSearch.onKeyDown(KeyEvent.KEYCODE_ENTER, KeyEvent(KeyEvent.ACTION_DOWN, KeyEvent.KEYCODE_ENTER))

        verify<ImageListPresenter>(presenter).onSearchClick("one")
    }

    private fun initImageInfoList() {
        imageInfoList = ArrayList()
        for (i in 0..49) {
            val imageInfo = ImageInfo()
            imageInfo.id = i
            imageInfo.previewURL = "https:\\\\previewurl$i"
            imageInfo.largeImageURL = "https:\\\\largeimagewurl$i"
            imageInfoList!!.add(imageInfo)
        }
    }
}